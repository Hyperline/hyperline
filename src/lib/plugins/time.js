import React, { Component } from 'react'
import moment from 'moment'
import SvgIcon from '../utils/svg-icon'

class PluginIcon extends Component {
  render() {
    return (
      <SvgIcon>
        <g fill="none" fillRule="evenodd">
          <g
            className='time-icon'
            transform="translate(1.000000, 1.000000)"
          >
            <g>
              <path d="M0,0 L14,0 L14,14 L0,14 L0,0 Z M1,1 L13,1 L13,13 L1,13 L1,1 Z" />
              <path d="M6,2 L7,2 L7,7 L6,7 L6,2 Z M6,7 L10,7 L10,8 L6,8 L6,7 Z" />
            </g>
          </g>
        </g>

        <style jsx>{`
          .time-icon {
            fill: #fff;
          }
        `}</style>
      </SvgIcon>
    )
  }
}

export default class Time extends Component {
  static displayName() {
    return 'time'
  }

  constructor(props) {
    super(props)

    this.state = {
      time: this.getCurrentTime()
    }
  }

  componentDidMount() {
    this.interval = setInterval(() => {
      this.setState({ time: this.getCurrentTime() })
    }, 100)
  }

  componentWillUnmount() {
    clearInterval(this.interval)
  }

  getCurrentTime() {
    // TODO: Allow for format overriding by the user
    return moment().format('LTS')
  }

  render() {
    return (
      <div className='wrapper'>
        <PluginIcon /> {this.state.time}

        <style jsx>{`
          .wrapper {
            display: flex;
            align-items: center;
          }
        `}</style>
      </div>
    )
  }
}
