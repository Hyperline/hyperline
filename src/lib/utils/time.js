import moment from 'moment'

export function formatUptime(uptime) {
  const uptimeInHours = Number((uptime/3600).toFixed(0));

  if (uptimeInHours === 0) {
    return '0h'
  }

  const uptimeInMoment = moment.duration(uptimeInHours, 'hours')
  const days = uptimeInMoment.days()
  const hours = uptimeInMoment.hours()
  const daysFormatted = days ? days + 'd' : ''
  const hoursFormatted = hours ? hours + 'h' : ''

  return [ daysFormatted, hoursFormatted ].filter(Boolean).join(' ')
}
