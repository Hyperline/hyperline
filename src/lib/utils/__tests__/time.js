/* eslint-env jest */
import { formatUptime } from '../time'

describe('time', () => {
  describe('formatUptime', () => {
    it('is a function', () => expect(formatUptime).toMatchSnapshot())
    it('3 seconds', () => expect(formatUptime(3)).toMatchSnapshot())
    it('23 minutes', () => expect(formatUptime(1380)).toMatchSnapshot())
    it('29 minutes', () => expect(formatUptime(1740)).toMatchSnapshot())
    it('30 minutes', () => expect(formatUptime(1800)).toMatchSnapshot())
    it('31 minutes', () => expect(formatUptime(1860)).toMatchSnapshot())
    it('3 hours 48 minutes', () => expect(formatUptime(13680)).toMatchSnapshot())
    it('4 hours', () => expect(formatUptime(14400)).toMatchSnapshot())
    it('23 hours', () => expect(formatUptime(82800)).toMatchSnapshot())
    it('23 hours 17 minutes', () => expect(formatUptime(83820)).toMatchSnapshot())
    it('23 hours 49 minutes', () => expect(formatUptime(85740)).toMatchSnapshot())
    it('1 day 4 hours', () => expect(formatUptime(100800)).toMatchSnapshot())
    it('3 days', () => expect(formatUptime(259200)).toMatchSnapshot())
    it('3 days 8 minutes', () => expect(formatUptime(259680)).toMatchSnapshot())
  })
})
