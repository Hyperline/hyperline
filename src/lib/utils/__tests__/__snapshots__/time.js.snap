// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`time formatUptime 1 day 4 hours 1`] = `"1d 4h"`;

exports[`time formatUptime 3 days 1`] = `"3d"`;

exports[`time formatUptime 3 days 8 minutes 1`] = `"3d"`;

exports[`time formatUptime 3 hours 48 minutes 1`] = `"4h"`;

exports[`time formatUptime 3 seconds 1`] = `"0h"`;

exports[`time formatUptime 4 hours 1`] = `"4h"`;

exports[`time formatUptime 23 hours 1`] = `"23h"`;

exports[`time formatUptime 23 hours 17 minutes 1`] = `"23h"`;

exports[`time formatUptime 23 hours 49 minutes 1`] = `"1d"`;

exports[`time formatUptime 23 minutes 1`] = `"0h"`;

exports[`time formatUptime 29 minutes 1`] = `"0h"`;

exports[`time formatUptime 30 minutes 1`] = `"1h"`;

exports[`time formatUptime 31 minutes 1`] = `"1h"`;

exports[`time formatUptime is a function 1`] = `[Function]`;
